File: study/examples/Algorithm/cpp/02_array_string.cpp
/*
 * 배열과 문자열 (Array and String)
 * Two Pointers, Sliding Window, Prefix Sum
 *
 * 배열과 문자열 처리의 핵심 기법들입니다.
 */

#include <iostream>
#include <vector>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <algorithm>
#include <climits>

using namespace std;

// =============================================================================
// 1. 투 포인터 (Two Pointers)
// =============================================================================

// 정렬된 배열에서 두 수의 합
pair<int, int> twoSumSorted(const vector<int>& arr, int target) {
    int left = 0, right = arr.size() - 1;

    while (left < right) {
        int sum = arr[left] + arr[right];
        if (sum == target)
            return {left, right};
        else if (sum < target)
            left++;
        else
            right--;
    }

    return {-1, -1};
}

// 물 담기 (Container With Most Water)
int maxArea(const vector<int>& height) {
    int left = 0, right = height.size() - 1;
    int maxWater = 0;

    while (left < right) {
        int h = min(height[left], height[right]);
        int w = right - left;
        maxWater = max(maxWater, h * w);

        if (height[left] < height[right])
            left++;
        else
            right--;
    }

    return maxWater;
}

// 세 수의 합
vector<vector<int>> threeSum(vector<int>& nums) {
    vector<vector<int>> result;
    sort(nums.begin(), nums.end());
    int n = nums.size();

    for (int i = 0; i < n - 2; i++) {
        if (i > 0 && nums[i] == nums[i - 1]) continue;

        int left = i + 1, right = n - 1;
        while (left < right) {
            int sum = nums[i] + nums[left] + nums[right];
            if (sum == 0) {
                result.push_back({nums[i], nums[left], nums[right]});
                while (left < right && nums[left] == nums[left + 1]) left++;
                while (left < right && nums[right] == nums[right - 1]) right--;
                left++;
                right--;
            } else if (sum < 0) {
                left++;
            } else {
                right--;
            }
        }
    }

    return result;
}

// =============================================================================
// 2. 슬라이딩 윈도우 (Sliding Window)
// =============================================================================

// 고정 크기 윈도우 최대 합
int maxSumWindow(const vector<int>& arr, int k) {
    int n = arr.size();
    if (n < k) return -1;

    int windowSum = 0;
    for (int i = 0; i < k; i++) {
        windowSum += arr[i];
    }

    int maxSum = windowSum;
    for (int i = k; i < n; i++) {
        windowSum += arr[i] - arr[i - k];
        maxSum = max(maxSum, windowSum);
    }

    return maxSum;
}

// 중복 없는 최장 부분 문자열
int lengthOfLongestSubstring(const string& s) {
    unordered_set<char> charSet;
    int maxLen = 0;
    int left = 0;

    for (int right = 0; right < (int)s.length(); right++) {
        while (charSet.count(s[right])) {
            charSet.erase(s[left]);
            left++;
        }
        charSet.insert(s[right]);
        maxLen = max(maxLen, right - left + 1);
    }

    return maxLen;
}

// 합이 target 이상인 최소 길이 부분 배열
int minSubArrayLen(int target, const vector<int>& nums) {
    int n = nums.size();
    int minLen = INT_MAX;
    int sum = 0;
    int left = 0;

    for (int right = 0; right < n; right++) {
        sum += nums[right];
        while (sum >= target) {
            minLen = min(minLen, right - left + 1);
            sum -= nums[left];
            left++;
        }
    }

    return minLen == INT_MAX ? 0 : minLen;
}

// =============================================================================
// 3. 프리픽스 합 (Prefix Sum)
// =============================================================================

class PrefixSum {
private:
    vector<long long> prefix;

public:
    PrefixSum(const vector<int>& arr) {
        int n = arr.size();
        prefix.resize(n + 1, 0);
        for (int i = 0; i < n; i++) {
            prefix[i + 1] = prefix[i] + arr[i];
        }
    }

    // [left, right] 구간 합
    long long rangeSum(int left, int right) {
        return prefix[right + 1] - prefix[left];
    }
};

// 합이 k인 부분 배열 개수
int subarraySum(const vector<int>& nums, int k) {
    unordered_map<int, int> prefixCount;
    prefixCount[0] = 1;
    int sum = 0;
    int count = 0;

    for (int num : nums) {
        sum += num;
        if (prefixCount.count(sum - k)) {
            count += prefixCount[sum - k];
        }
        prefixCount[sum]++;
    }

    return count;
}

// =============================================================================
// 4. 카데인 알고리즘 (Kadane's Algorithm)
// =============================================================================

int maxSubArray(const vector<int>& nums) {
    int maxSum = nums[0];
    int currentSum = nums[0];

    for (size_t i = 1; i < nums.size(); i++) {
        currentSum = max(nums[i], currentSum + nums[i]);
        maxSum = max(maxSum, currentSum);
    }

    return maxSum;
}

// 최대 부분 배열의 인덱스도 반환
tuple<int, int, int> maxSubArrayWithIndices(const vector<int>& nums) {
    int maxSum = nums[0];
    int currentSum = nums[0];
    int start = 0, end = 0, tempStart = 0;

    for (size_t i = 1; i < nums.size(); i++) {
        if (nums[i] > currentSum + nums[i]) {
            currentSum = nums[i];
            tempStart = i;
        } else {
            currentSum += nums[i];
        }

        if (currentSum > maxSum) {
            maxSum = currentSum;
            start = tempStart;
            end = i;
        }
    }

    return {maxSum, start, end};
}

// =============================================================================
// 5. 문자열 처리
// =============================================================================

// 팰린드롬 체크
bool isPalindrome(const string& s) {
    int left = 0, right = s.length() - 1;
    while (left < right) {
        if (s[left] != s[right]) return false;
        left++;
        right--;
    }
    return true;
}

// 애너그램 체크
bool isAnagram(const string& s, const string& t) {
    if (s.length() != t.length()) return false;

    vector<int> count(26, 0);
    for (size_t i = 0; i < s.length(); i++) {
        count[s[i] - 'a']++;
        count[t[i] - 'a']--;
    }

    for (int c : count) {
        if (c != 0) return false;
    }
    return true;
}

// 문자열 압축
string compress(const string& s) {
    if (s.empty()) return s;

    string result;
    int count = 1;

    for (size_t i = 1; i <= s.length(); i++) {
        if (i < s.length() && s[i] == s[i - 1]) {
            count++;
        } else {
            result += s[i - 1];
            if (count > 1) {
                result += to_string(count);
            }
            count = 1;
        }
    }

    return result.length() < s.length() ? result : s;
}

// =============================================================================
// 테스트
// =============================================================================

int main() {
    cout << "============================================================" << endl;
    cout << "배열과 문자열 예제" << endl;
    cout << "============================================================" << endl;

    // 1. 투 포인터
    cout << "\n[1] 투 포인터" << endl;
    vector<int> sorted = {1, 2, 3, 4, 6};
    auto [idx1, idx2] = twoSumSorted(sorted, 6);
    cout << "    배열: [1,2,3,4,6], 타겟: 6" << endl;
    cout << "    인덱스: (" << idx1 << ", " << idx2 << ")" << endl;

    vector<int> heights = {1, 8, 6, 2, 5, 4, 8, 3, 7};
    cout << "    물 담기: " << maxArea(heights) << endl;

    // 2. 슬라이딩 윈도우
    cout << "\n[2] 슬라이딩 윈도우" << endl;
    vector<int> arr = {1, 4, 2, 10, 23, 3, 1, 0, 20};
    cout << "    크기 4 윈도우 최대 합: " << maxSumWindow(arr, 4) << endl;

    cout << "    \"abcabcbb\" 최장 부분 문자열: " << lengthOfLongestSubstring("abcabcbb") << endl;

    vector<int> nums = {2, 3, 1, 2, 4, 3};
    cout << "    합 >= 7 최소 길이: " << minSubArrayLen(7, nums) << endl;

    // 3. 프리픽스 합
    cout << "\n[3] 프리픽스 합" << endl;
    vector<int> prefixArr = {1, 2, 3, 4, 5};
    PrefixSum ps(prefixArr);
    cout << "    배열: [1,2,3,4,5]" << endl;
    cout << "    [1,3] 구간 합: " << ps.rangeSum(1, 3) << endl;

    vector<int> subarr = {1, 1, 1};
    cout << "    합이 2인 부분 배열 개수: " << subarraySum(subarr, 2) << endl;

    // 4. 카데인 알고리즘
    cout << "\n[4] 카데인 알고리즘" << endl;
    vector<int> kadane = {-2, 1, -3, 4, -1, 2, 1, -5, 4};
    cout << "    배열: [-2,1,-3,4,-1,2,1,-5,4]" << endl;
    cout << "    최대 부분 배열 합: " << maxSubArray(kadane) << endl;

    // 5. 문자열 처리
    cout << "\n[5] 문자열 처리" << endl;
    cout << "    \"racecar\" 팰린드롬: " << (isPalindrome("racecar") ? "예" : "아니오") << endl;
    cout << "    \"anagram\"/\"nagaram\" 애너그램: " << (isAnagram("anagram", "nagaram") ? "예" : "아니오") << endl;
    cout << "    \"aabcccccaaa\" 압축: " << compress("aabcccccaaa") << endl;

    // 6. 복잡도 요약
    cout << "\n[6] 복잡도 요약" << endl;
    cout << "    | 알고리즘        | 시간복잡도 |" << endl;
    cout << "    |-----------------|------------|" << endl;
    cout << "    | 투 포인터       | O(n)       |" << endl;
    cout << "    | 슬라이딩 윈도우 | O(n)       |" << endl;
    cout << "    | 프리픽스 합     | O(1) 쿼리  |" << endl;
    cout << "    | 카데인          | O(n)       |" << endl;

    cout << "\n============================================================" << endl;

    return 0;
}

File: study/examples/Algorithm/python/02_array_string.py
"""
투 포인터 (Two Pointer) 기법
Two Pointer Technique

두 개의 포인터를 사용하여 배열/리스트 문제를 효율적으로 해결합니다.
"""

from typing import List, Tuple, Optional


# =============================================================================
# 1. 두 수의 합 (정렬된 배열)
# =============================================================================
def two_sum_sorted(arr: List[int], target: int) -> Optional[Tuple[int, int]]:
    """
    정렬된 배열에서 합이 target인 두 수의 인덱스 찾기
    시간복잡도: O(n), 공간복잡도: O(1)
    """
    left, right = 0, len(arr) - 1

    while left < right:
        current_sum = arr[left] + arr[right]

        if current_sum == target:
            return (left, right)
        elif current_sum < target:
            left += 1  # 합이 작으면 왼쪽 포인터를 오른쪽으로
        else:
            right -= 1  # 합이 크면 오른쪽 포인터를 왼쪽으로

    return None


# =============================================================================
# 2. 세 수의 합 (3Sum)
# =============================================================================
def three_sum(arr: List[int]) -> List[List[int]]:
    """
    합이 0인 세 수의 조합 모두 찾기 (중복 제거)
    시간복잡도: O(n²), 공간복잡도: O(1)
    """
    arr.sort()
    result = []
    n = len(arr)

    for i in range(n - 2):
        # 중복 건너뛰기
        if i > 0 and arr[i] == arr[i - 1]:
            continue

        left, right = i + 1, n - 1

        while left < right:
            total = arr[i] + arr[left] + arr[right]

            if total == 0:
                result.append([arr[i], arr[left], arr[right]])
                # 중복 건너뛰기
                while left < right and arr[left] == arr[left + 1]:
                    left += 1
                while left < right and arr[right] == arr[right - 1]:
                    right -= 1
                left += 1
                right -= 1
            elif total < 0:
                left += 1
            else:
                right -= 1

    return result


# =============================================================================
# 3. 물 담기 (Container With Most Water)
# =============================================================================
def max_water(heights: List[int]) -> int:
    """
    두 벽 사이에 담을 수 있는 최대 물의 양
    시간복잡도: O(n), 공간복잡도: O(1)
    """
    left, right = 0, len(heights) - 1
    max_area = 0

    while left < right:
        # 현재 면적 계산
        width = right - left
        height = min(heights[left], heights[right])
        area = width * height
        max_area = max(max_area, area)

        # 더 낮은 쪽의 포인터 이동
        if heights[left] < heights[right]:
            left += 1
        else:
            right -= 1

    return max_area


# =============================================================================
# 4. 정렬된 두 배열 합병
# =============================================================================
def merge_sorted_arrays(arr1: List[int], arr2: List[int]) -> List[int]:
    """
    정렬된 두 배열을 하나의 정렬된 배열로 합병
    시간복잡도: O(n + m), 공간복잡도: O(n + m)
    """
    result = []
    i, j = 0, 0

    while i < len(arr1) and j < len(arr2):
        if arr1[i] <= arr2[j]:
            result.append(arr1[i])
            i += 1
        else:
            result.append(arr2[j])
            j += 1

    # 남은 요소 추가
    result.extend(arr1[i:])
    result.extend(arr2[j:])

    return result


# =============================================================================
# 5. 중복 제거 (정렬된 배열)
# =============================================================================
def remove_duplicates(arr: List[int]) -> int:
    """
    정렬된 배열에서 중복 제거 (in-place)
    반환값: 고유한 요소의 개수
    시간복잡도: O(n), 공간복잡도: O(1)
    """
    if not arr:
        return 0

    write_idx = 1  # 다음 고유 값을 쓸 위치

    for read_idx in range(1, len(arr)):
        if arr[read_idx] != arr[write_idx - 1]:
            arr[write_idx] = arr[read_idx]
            write_idx += 1

    return write_idx


# =============================================================================
# 6. 회문 검사 (Palindrome)
# =============================================================================
def is_palindrome(s: str) -> bool:
    """
    문자열이 회문인지 검사 (알파벳/숫자만 비교)
    시간복잡도: O(n), 공간복잡도: O(1)
    """
    left, right = 0, len(s) - 1

    while left < right:
        # 알파벳/숫자가 아니면 건너뛰기
        while left < right and not s[left].isalnum():
            left += 1
        while left < right and not s[right].isalnum():
            right -= 1

        if s[left].lower() != s[right].lower():
            return False

        left += 1
        right -= 1

    return True


# =============================================================================
# 7. 슬라이딩 윈도우 (최대 합)
# =============================================================================
def max_sum_subarray(arr: List[int], k: int) -> int:
    """
    크기 k인 연속 부분 배열의 최대 합
    시간복잡도: O(n), 공간복잡도: O(1)
    """
    if len(arr) < k:
        return 0

    # 초기 윈도우 합
    window_sum = sum(arr[:k])
    max_sum = window_sum

    # 윈도우 슬라이딩
    for i in range(k, len(arr)):
        window_sum = window_sum - arr[i - k] + arr[i]
        max_sum = max(max_sum, window_sum)

    return max_sum


# =============================================================================
# 테스트
# =============================================================================
def main():
    print("=" * 60)
    print("투 포인터 (Two Pointer) 기법 예제")
    print("=" * 60)

    # 1. 두 수의 합
    print("\n[1] 두 수의 합 (정렬된 배열)")
    arr = [1, 2, 3, 4, 5, 6, 7, 8, 9]
    target = 10
    result = two_sum_sorted(arr, target)
    print(f"    배열: {arr}")
    print(f"    타겟: {target}")
    print(f"    결과: 인덱스 {result} -> {arr[result[0]]} + {arr[result[1]]} = {target}")

    # 2. 세 수의 합
    print("\n[2] 세 수의 합 (3Sum)")
    arr = [-1, 0, 1, 2, -1, -4]
    result = three_sum(arr)
    print(f"    배열: {arr}")
    print(f"    합이 0인 조합: {result}")

    # 3. 물 담기
    print("\n[3] 물 담기 (Container With Most Water)")
    heights = [1, 8, 6, 2, 5, 4, 8, 3, 7]
    result = max_water(heights)
    print(f"    높이: {heights}")
    print(f"    최대 물의 양: {result}")

    # 4. 정렬된 배열 합병
    print("\n[4] 정렬된 두 배열 합병")
    arr1 = [1, 3, 5, 7]
    arr2 = [2, 4, 6, 8, 10]
    result = merge_sorted_arrays(arr1, arr2)
    print(f"    배열1: {arr1}")
    print(f"    배열2: {arr2}")
    print(f"    합병 결과: {result}")

    # 5. 중복 제거
    print("\n[5] 중복 제거 (in-place)")
    arr = [1, 1, 2, 2, 2, 3, 4, 4, 5]
    count = remove_duplicates(arr)
    print(f"    원본: [1, 1, 2, 2, 2, 3, 4, 4, 5]")
    print(f"    고유 요소 개수: {count}")
    print(f"    결과 배열 (앞 {count}개): {arr[:count]}")

    # 6. 회문 검사
    print("\n[6] 회문 검사")
    test_strings = ["A man, a plan, a canal: Panama", "race a car", "Was it a car or a cat I saw?"]
    for s in test_strings:
        result = is_palindrome(s)
        print(f"    '{s}' -> {result}")

    # 7. 슬라이딩 윈도우
    print("\n[7] 슬라이딩 윈도우 (크기 k 최대 합)")
    arr = [2, 1, 5, 1, 3, 2]
    k = 3
    result = max_sum_subarray(arr, k)
    print(f"    배열: {arr}, k={k}")
    print(f"    최대 합: {result}")

    print("\n" + "=" * 60)


if __name__ == "__main__":
    main()
